File: src/main/java/org/algohub/engine/collection/BinaryTreeNode.java
package org.algohub.engine.collection;

import java.util.LinkedList;
import java.util.Queue;

/** Binary Tree Node. */
@SuppressWarnings({"PMD.BeanMembersShouldSerialize", "PMD.CommentRequired", "PMD.ShortVariable"})
public class BinaryTreeNode<E> {
  public E value;
  public BinaryTreeNode<E> left;
  public BinaryTreeNode<E> right;

  /**
   * has a left child, but it's a null node.
   */
  public transient boolean leftIsNull;
  /**
   * has a right child, but it's a null node.
   */
  public transient boolean rightIsNull;
  //public transient BinaryTreeNode<E> next;

  /**
   * Constructs an empty binary tree.
   */
  public BinaryTreeNode() {
    value = null;
    left = null;
    right = null;
  }

  /**
   * Constructs an binary tree with one element.
   */
  public BinaryTreeNode(final E value) {
    this.value = value;
    left = null;
    right = null;
  }

  private static boolean isSameTree(final BinaryTreeNode p, final BinaryTreeNode q) {
    if (p == null && q == null) {
      return true;
    }
    if (p == null || q == null) {
      return false;
    }

    return p.value.equals(q.value) && isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
  }

  // get the insertion position
  @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
  private NodeAndFather tail() {
    final Queue<NodeAndFather> queue = new LinkedList<>();
    if (value != null) {
      queue.add(new NodeAndFather(this, null, true));
    }

    while (!queue.isEmpty()) {
      final NodeAndFather nodeAndFather = queue.poll();

      if (nodeAndFather.node != null) {
        if (nodeAndFather.node.left != null || nodeAndFather.node.leftIsNull) {
          queue.add(new NodeAndFather(nodeAndFather.node.left, nodeAndFather.node, false));
        } else {
          return new NodeAndFather(nodeAndFather.node.left, nodeAndFather.node, false);
        }
        if (nodeAndFather.node.right != null || nodeAndFather.node.rightIsNull) {
          queue.add(new NodeAndFather(nodeAndFather.node.right, nodeAndFather.node, true));
        } else {
          return new NodeAndFather(nodeAndFather.node.right, nodeAndFather.node, true);
        }
      }
    }
    return null;
  }

  /**
   * return next insert point.
   */
  public void add(final E value) {
    final NodeAndFather last = tail();
    if (last == null) {
      this.value = value;
      return;
    }

    if (last.isRight) {
      if (value == null) {
        last.father.rightIsNull = true;
      } else {
        last.father.right = new BinaryTreeNode<>(value);
      }
    } else {
      if (value == null) {
        last.father.leftIsNull = true;
      } else {
        last.father.left = new BinaryTreeNode<>(value);
      }
    }
  }

  @Override public boolean equals(final Object obj) {
    if (obj == this) {
      return true;
    }
    if (!(obj instanceof BinaryTreeNode)) {
      return false;
    }

    final BinaryTreeNode<E> p = this;
    final BinaryTreeNode<E> q = (BinaryTreeNode<E>) obj;
    return isSameTree(p, q);
  }

  @Override public String toString() {
    if (value == null) {
      return "[]";
    }

    final StringBuilder sb = new StringBuilder();
    sb.append('[');
    final Queue<BinaryTreeNode<E>> queue = new LinkedList<>();
    queue.add(this);
    while (!queue.isEmpty()) {
      final BinaryTreeNode<E> node = queue.poll();
      sb.append(',').append(node.value == null ? "null" : node.value.toString());
    }

    return sb.toString();
  }

  @Override public int hashCode() {
    int hashCode = 1;
    final Queue<BinaryTreeNode<E>> queue = new LinkedList<>();
    queue.add(this);

    while (!queue.isEmpty()) {
      final BinaryTreeNode<E> node = queue.poll();
      hashCode = 31 * hashCode + (node.value == null ? 0 : node.value.hashCode());

      if (node.left != null) {
        queue.offer(node.left);
      }
      if (node.right != null) {
        queue.offer(node.right);
      }
    }

    return hashCode;
  }


  private static class NodeAndFather {
    final private BinaryTreeNode node;
    final private BinaryTreeNode father;
    final private boolean isRight;  // node is the right child

    public NodeAndFather(final BinaryTreeNode node, final BinaryTreeNode father, final boolean isRight) {
      this.node = node;
      this.father = father;
      this.isRight = isRight;
    }
  }
}

File: src/main/resources/python/algohub/binary_tree_node.py
from .common_equality_mixin import CommonEqualityMixin
from collections import deque

class BinaryTreeNode(CommonEqualityMixin):
    def __init__(self, value=None, left=None, right=None, left_is_null=False,
                 right_is_null=False):
        self.value = value
        self.left = left # type: BinaryTreeNode
        self.right = right # type: BinaryTreeNode
        self.left_is_null = left_is_null
        self.right_is_null = right_is_null

    class NodeAndFather(object):
        def __init__(self, node, father, is_right):
            self.node = node # type: BinaryTreeNode
            self.father = father # type: BinaryTreeNode
            self.is_right = is_right

    def __tail(self):
        queue = deque()  # type: deque[BinaryTreeNode.NodeAndFather]
        if self.value is not None:
            queue.append(self.NodeAndFather(self, None, True))

        while len(queue) > 0:
            pos = queue.popleft()

            if pos.node is not None:
                if pos.node.left is not None or pos.node.left_is_null:
                    queue.append(self.NodeAndFather(pos.node.left, pos.node, False))
                else:
                    return self.NodeAndFather(pos.node.left, pos.node, False)
                if pos.node.right is not None or pos.node.right_is_null:
                    queue.append(self.NodeAndFather(pos.node.right, pos.node, True))
                else:
                    return self.NodeAndFather(pos.node.right, pos.node, True)
        return None

    def add(self, value):
        last = self.__tail()
        if last is None:
            self.value = value
            return

        if last.is_right:
            if value is not None:
                last.father.right = BinaryTreeNode(value)
            else:
                last.father.right_is_null = True
        else:
            if value is not None:
                last.father.left = BinaryTreeNode(value)
            else:
                last.father.left_is_null = True

    def __str__(self):
        result = '['
        current = deque()
        next = deque()
        if self.value is not None:
            current.append(self)

        while len(current) > 0:
            level = []

            while len(current) > 0:
                level.append(current.popleft())

            last_not_null_index = -1
            for i in reversed(range(len(level))):
                if level[i] is not None:
                    last_not_null_index = i
                    break

            for i in range(last_not_null_index + 1):
                node = level[i]
                if node is not None:
                    result.append(str(node.value) + ',')
                else:
                    result.append('null,')

                if node is not None:
                    next.append(node.left)
                    next.append(node.right)

            # swap current and next
            tmp = current
            current = next
            next = tmp

        return result[:-1] + ']'

